README
1. Usage
========

  hwinfo [debug=deb_flag] [log=log_file] [list[+]=hw_item] [[+|-]probe_option1] [[+|-]probe_option2] ...

Examples:

  hwinfo
    - probes for nearly everything

  hwinfo +all
    - probes for everything

  hwinfo log=hw_log
    - default probing, output is written to hw_log
      *** Please don't use "hwinfo >some_log 2>&1" to store the output into a
      log file! ***

  hwinfo -all +ide
    - probe for ide drives only

  hwinfo +all -isapnp -floppy
    - probe for everything *except* isapnp and floppies

Some probing flags do not stand for complete modules but enable additional
features; e.g. "pci.range" or "cdrom.info".

Example:
  hwinfo -all +cdrom
  gives a list of all cdrom's

  hwinfo -all +cdrom.info
  additionally reads the iso9660 info

To get a list of all supported probing flags, call 'hwinfo -all'.


2. Environment variables/kernel cmdline parameters
==================================================

You can control the hardware probing using the environment variable
"hwprobe" and the kernel cmdline parameters "hwprobe" & "x11".

If "hwprobe" is set on the kernel cmdline, the environment variable
"hwprobe" is ignored. Otherwise, the meaning of both is exactly the same.

  x11
    - Controls the X server/X version to prefer. It is either "3" or "4" or the
      name of the X server/X server module to use (XF3: server name without
      the leading "XF86_", XF4: the server module name).

      Examples:
        x11=3
          o prefer XFree86 3.x over 4.x (if possible)

        x11=SVGA
          o use the XF86_SVGA server (*** Even if the server does not
            support the card! ***)

        x11=nv
          o use the XF4 nv_drv.o driver (*** Even if the module does not
            support the card! ***)

  hwprobe
    - Controls which probing flags should *always* be set/cleared. (These
      settings *cannot* be overridden by command line switches.)

      Examples:
        hwprobe=-isapnp
          o *never* do any isapnp probing

        hwprobe=-braille,-modem
          o don't look for braille displays & modems

    - Allows you to add and remove hardware from the probing results. In this case
      the syntax is (-: remove, +: add at end of list, <nothing>: add at begin of
      list):
        hwprobe=[+-]<device_class>:<vendor_id>:<device_id>[:<unix_device_file>]

      <device_class>, <vendor_id> and <device_id> are device ids as used by libhd.
      See the output of hwinfo for examples. In connection with "-", you can use
      "*" as a placeholder that matches every id.

      Note: <unix_device_file> is optional.

      Note2: you cannot usefully *add* hardware that needs more info than that
      given by the hwprobe entry. Disks & floppies are notable examples.
      (But you can *remove* them.)

      Here is a typical hwinfo output for a mouse, with the relevant ids
      underlined (<device_class> is the combined base_class & sub_class),
      [see 1st example below]:

        14: PS/2 00.0: 10500 PS/2 Mouse
                       ^^^^^ -->	<device_class>
          [Created at mouse.110]
          Vendor: s0200 "Unknown"
                  ^^^^^  -->		<vendor_id>
          Model: 0002 "Generic PS/2 Mouse"
                 ^^^^  -->		<device_id>
          Device File: /dev/psaux
                       ^^^^^^^^^^ -->	<unix_device_file>
          Driver Info #0:
            XFree86 Protocol: ps/2
            GPM Protocol: ps2
          Attached to: #8 (PS/2 Controller)


      Examples:
        hwprobe=+10500:s200:2:/dev/psaux
          o add a ps/2 mouse [at the end of the hardware list]

        hwprobe=10500:s200:2:/dev/psaux
          o add a ps/2 mouse [at the start of the hardware list, so it
            is our default mouse]

        hwprobe=+10b00:s5001:0:/dev/ttyS0
          o add a braille display connected to /dev/ttyS0

        hwprobe=-10500:s200:2:/dev/psaux
          o remove ps/2 mice attached to /dev/psaux

        hwprobe=-10500:s200:2
          o remove all ps/2 mice

        hwprobe=-10500:*:*
          o remove all ps/2 mice

        hwprobe=-*:*:*:/dev/hdc
          o remove /dev/hdc

        hwprobe=+401:1274:5000
          o add an ensoniq sound card

      Grafics cards are are slightly trickier:

        hwprobe=+300:1014:b7
          o add a Fire GL1 card
            Note: this way you'll get a multihead config. You'll probably
            rather want the following example.

        hwprobe=-300:*:*,+300:1014:b7
          o remove all graphics cards; then add a Fire GL1 card

        hwprobe=+400:121a:1
          o add a 3fx voodoo card (Note the class "400", not "300"!)

    - for more ids, see "src/ids/names.*" and "src/ids/drivers.*"



ChangeLog
23/1/2006:	v11.35
	- better monitor detection
	- updated X11 dat
	- qla* modules require qlogic-firmware

16/1/2006:	v11.34
	- added qla4xxx workaround (#141069)

9/1/2006:	v11.33
	- adjusted i2o driver info (#129301)

9/1/2006:	v11.32
	- jg: wlan update

9/12/2005:	v11.31
	- add VirtualIron network cards (#135309)
	- updated X11 data

7/12/2005:	v11.30
	- report ide-generic for non-pci ide interfaces

6/12/2005:	v11.29
	- s390: make sysfs reading more robust
	- s390: fixed stack corruption on exotic hw setups (bug #128453)
	- s390: detect OSN devices as QETH interfaces (bug #120724)
	- s390: add module entry for P/390 LCS device (bug #81207)
	- adjust input device detection to new sysfs layout (#134032)

28/10/2005:	v11.28
	- added Dell Notebook LCD size (#130180)
	- another Notebook added (#130208)

14/10/2005:	v11.27
	- find xen ethernet cards
	- remove old hardware entries (#118673)
	- better support for virtual i/o hardware on ppc (#119592)
	- added r8180 to wlan list (#115268)
	- jg: fix wlan includes

20/9/2005:	v11.26
	- don't assume mp records start below 1MB (#112699)
	- fix unaligned accesses (#97838)
	- gcc 4.1 fixes

14/9/2005:	v11.25
	- fix forcedeth, again (#116394)

13/9/2005:	v11.24
	- better pcmcia module detection (#104517)

12/9/2005:	v11.23
	- update forcedeth driver info (#116394)

7/9/2005:	v11.22
	- fix vmware check (broke in UML) (#115424)
	- toshiba notebooks need ial (#113714)

6/9/2005:	v11.21
	- toshiba notebooks don't need fnfx (#113714)

5/9/2005:	v11.20
	- ppc: read 'LCD,EDID', too (#115085)
	- module renamed: skystar2 -> b2c2-flexcop-pci (#115262)
	- dto, dmasound -> snd-powermac (#112919)

2/9/2005:	v11.19
	- report more ivtv cards, and make them (analog) tv cards (#113195)

1/9/2005:	v11.18
	- fixed udev parser bug (#113766)
	- use new 'udevinfo -e' option (#113766)
	- add PowerBook LCD to db (#113795)
	- cardbus device detection fixed (#113708)

30/8/2005:	v11.17
	- jg: added support for WPA Wireless Extension in WLAN probing
	  (fixes faulty probing of ipw2x00 and hostap devices)
	- jg: added rt2570 & zd1201 driver to list of WLAN adapters (#105623)

29/8/2005:	v11.16
	- fix alps touchpad detection (#98947, #103075)
	- drop ltmodem support (#113336)
	- don't make too many floppies (#113571)
	- updated X11 data
	- updated pci ids
	- added WinTV PVR-350 as dvb card

23/8/2005:	v11.15
	- fix pppoe detection (#106836)

22/8/2005:	v11.14
	- fix pcmcia controller detection

19/8/2005:	v11.13
	- rip out all special tulip-related module info (#105730)
	- load lp module (#104874)
	- pcmcia probing works again (#103740)
	- support 'modalias' sysfs entries (#103740, #103978)
	- getsysinfo collects a bit more info
	- drop pci.handmap code (after talking to zoz)

16/8/2005:	v11.12
	- fixed evil bug when hal is running (#104654)

10/8/2005:	v11.11
	- added hdtv cards (#102933)
	- find input device udi (#102575)
	- fixed usb device udi matching (#102575)
	- read modules.alias, not modules.pcimap
	- updated X11 data

8/8/2005:	v11.10
	- drop special ADB input device scanning (#98324)
	- internal db function rework continued (#102575)
	- add udi when possible (#102575)
	- updated X11 data

1/8/2005:	v11.9
	- updated X11 data
	- implement compat wrapper to keep old db functions working
	- kkeil: AVM now suports 64 bit drivers

25/7/2005:	v11.8
	- updated X11 data
	- static mpt fusion pci id data removed (#97665)

21/7/2005:	v11.7
	- read/write udi-based persitent properties works

18/7/2005:	v11.6
	- rip out remaining old db access methods

6/7/2005:	v11.5
	- pci devs basically work
	- use giant ibm notebook database
	- added 'hwclass' entry to db format

5/7/2005:	v11.4
	- make C++ happy

4/7/2005:	v11.3
	- removed special ppc driver info (#91437)
	- skip Video BIOS checksum test
	- move libhd back to /usr
	- use hal/dbus

9/6/2005:	v11.2
	- getsysinfo collects more info
	- some more ibm notebooks
	- don't use de4x5
	- acpi: always load thermal & fan modules (#72146)

11/4/2005:	v11.1
	- smbios detection: don't assume dmi records start below 1MB
	- fixed kernel header include

1/4/2005:	v11.0
	- removed hwscan*
	- make it build with gcc4

22/3/2005:	v10.16
	- updated X11 data (#74130)

21/3/2005:	v10.15
	- mls: faster hwscand

21/3/2005:	v10.14
	- updated X11 data (#74022)

19/3/2005:	v10.13
	- fix wrong entry for Fritz!Card DSL SL USB and
	  make ppp mode detection safer (#71995)
	- adjusted ltmodem info (#71979)

18/3/2005:	v10.12
	- updated nongpl info
	- reverting slamr patch (#72301)
	- pcmcia info updated (#73057)
	- pci dev names updated
	- read /proc/modules less often
	- updated X11 data
	- aic7xxx/aic79xx driver info update
	- fix Fritz!Box Fon entry in database

17/3/2005:	v10.11
	- another notebook (#67412)
	- do wlan detection for --netcard (#73077)
	- remove Eicon cards from 64 bit archs, the driver does not support
	  64 bit yet (#55722)

16/3/2005:	v10.10
	- updated notebook display data
	- fix udev db reading
	- updated X11 data
	- ltmodem device id update
	- no check for gfx card changes (#72907)

14/3/2005:	v10.9
	- hardware detection for ISDN PCMCIA cards (#71208)
	- updated ISDN data
	- updated X11 data
	- better network type detection (#70991, #71231, #67325)
	- load some acpi modules on hp notebook (#72146)
	- fixed isapnp sound card detection (#67303)
	- slamr: ttySL0 -> ttyLT0 (#72301)

7/3/2005:	v10.8
	- updated X11 data

4/3/2005:	v10.7
	- s390: made ccwgroup walking more robust
	- fixed bios base mem detection
	- allow device class in 'hwinfo --db' request
	- updated X11 data

28/2/2005:	v10.6
	- jg: improved wlan patch (no dependency on libiw)

28/2/2005:	v10.5
	- fixed model name for some SGI hardware (#63973)
	- fix Fritz!Card DSL SL USB ID (#66674)
	- better heuristics to assign BIOS driver numbers (#66669)
	- removed ahci vs. ata_piix hack (#65218)

25/2/2005:	v10.4
	- report pci class 0x403 as audio devices (#66466)
	- load ahci *and* ata_piix (#65218)
	- symlink /dev/fb might be missing; trying fb0, too (#66501)

hwscan.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

#include "hd.h"
#include "hd_int.h"

struct option options[] = {
  { "help", 0, NULL, 'h' },
  { "verbose", 0, NULL, 'v' },
  { "version", 0, NULL, 400 },
  { "show", 1, NULL, 500 },
  { "list", 0, NULL, 501 },
  { "cfg", 1, NULL, 502 },
  { "avail", 1, NULL, 503 },
  { "need", 1, NULL, 504 },
  { "new", 0, NULL, 505 },
  { "fast", 0, NULL, 506 },
  { "silent", 0, NULL, 507 },
  { "boot", 0, NULL, 508 },
  { "active", 1, NULL, 509 },
  { "only", 1, NULL, 510 },
  { "sys", 0, NULL, 1000 + hw_sys },
  { "cpu", 0, NULL, 1000 + hw_cpu },
  { "keyboard", 0, NULL, 1000 + hw_keyboard },
  { "braille", 0, NULL, 1000 + hw_braille },
  { "mouse", 0, NULL, 1000 + hw_mouse },
  { "joystick", 0, NULL, 1000 + hw_joystick },
  { "printer", 0, NULL, 1000 + hw_printer },
  { "scanner", 0, NULL, 1000 + hw_scanner },
  { "chipcard", 0, NULL, 1000 + hw_chipcard },
  { "monitor", 0, NULL, 1000 + hw_monitor },
  { "tv", 0, NULL, 1000 + hw_tv },
  { "gfxcard", 0, NULL, 1000 + hw_display },
  { "framebuffer", 0, NULL, 1000 + hw_framebuffer },
  { "camera", 0, NULL, 1000 + hw_camera },
  { "sound", 0, NULL, 1000 + hw_sound },
  { "storage-ctrl", 0, NULL, 1000 + hw_storage_ctrl },
  { "storage_ctrl", 0, NULL, 1000 + hw_storage_ctrl },
  { "netcard", 0, NULL, 1000 + hw_network_ctrl },
  { "network-ctrl", 0, NULL, 1000 + hw_network_ctrl },
  { "network_ctrl", 0, NULL, 1000 + hw_network_ctrl },
  { "isdn", 0, NULL, 1000 + hw_isdn },
  { "modem", 0, NULL, 1000 + hw_modem },
  { "network", 0, NULL, 1000 + hw_network },
  { "disk", 0, NULL, 1000 + hw_disk },
  { "partition", 0, NULL, 1000 + hw_partition },
  { "cdrom", 0, NULL, 1000 + hw_cdrom },
  { "floppy", 0, NULL, 1000 + hw_floppy },
  { "update", 0, NULL, 1000 + hw_manual },
  { "usb-ctrl", 0, NULL, 1000 + hw_usb_ctrl },
  { "usb_ctrl", 0, NULL, 1000 + hw_usb_ctrl },
  { "usb", 0, NULL, 1000 + hw_usb },
  { "bios", 0, NULL, 1000 + hw_bios },
  { "pci", 0, NULL, 1000 + hw_pci },
  { "isapnp", 0, NULL, 1000 + hw_isapnp },
  { "bridge", 0, NULL, 1000 + hw_bridge },
  { "hub", 0, NULL, 1000 + hw_hub },
  { "scsi", 0, NULL, 1000 + hw_scsi },
  { "ide", 0, NULL, 1000 + hw_ide },
  { "memory", 0, NULL, 1000 + hw_memory },
  { "dvb", 0, NULL, 1000 + hw_dvb },
  { "pcmcia", 0, NULL, 1000 + hw_pcmcia },
  { "pcmcia_ctrl", 0, NULL, 1000 + hw_pcmcia_ctrl },
  { "ieee1394", 0, NULL, 1000 + hw_ieee1394 },
  { "firewire", 0, NULL, 1000 + hw_ieee1394 },
  { "ieee1394_ctrl", 0, NULL, 1000 + hw_ieee1394_ctrl },
  { "firewire_ctrl", 0, NULL, 1000 + hw_ieee1394_ctrl },
  { "hotplug", 0, NULL, 1000 + hw_hotplug },
  { "hotplug_ctrl", 0, NULL, 1000 + hw_hotplug_ctrl },
  { "zip", 0, NULL, 1000 + hw_zip },
  { "pppoe", 0, NULL, 1000 + hw_pppoe },
  { "dsl", 0, NULL, 1000 + hw_dsl },
  { "wlan", 0, NULL, 1000 + hw_wlan },
  { "block", 0, NULL, 1000 + hw_block },
  { "tape", 0, NULL, 1000 + hw_tape },
  { "vbe", 0, NULL, 1000 + hw_vbe },
  { "all", 0, NULL, 1000 + hw_all },
  { }
};

int verbose = 0;
hd_hw_item_t scan_item[100] = { };
unsigned scan_items = 0;
int found_items = 0;

struct {
  unsigned show:1;
  unsigned scan:1;
  unsigned list:1;
  unsigned config_cfg:1;
  unsigned config_avail:1;
  unsigned config_need:1;
  unsigned config_active:1;
  unsigned new:1;
  unsigned fast:1;
  unsigned silent:1;
  unsigned boot:1;
  str_list_t *only;
} opt;

void help(void);
int do_scan(hd_hw_item_t *items);
int do_show(char *id);
int do_list(hd_hw_item_t *items);
int do_config(int type, char *val, char *id);
int fast_ok(hd_hw_item_t *items);
int has_item(hd_hw_item_t *items, hd_hw_item_t item);
int has_hw_class(hd_t *hd, hd_hw_item_t *items);


int main(int argc, char **argv)
{
  int rc = 0;

#ifndef LIBHD_TINY

  char *id = NULL;
  char *config_cfg = NULL;
  char *config_avail = NULL;
  char *config_need = NULL;
  char *config_active = NULL;
  int i;
  int ok = 0;
  FILE *f;

  opterr = 0;

  while((i = getopt_long(argc, argv, "hv", options, NULL)) != -1) {
    switch(i) {
      case 'v':
        verbose++;
        break;

      case 400:
        printf("%s\n", hd_version());
        ok = 1;
        break;

      case 500:
        opt.show = 1;
        id = optarg;
        break;

      case 501:
        opt.list = 1;
        break;

      case 502:
        opt.config_cfg = 1;
        config_cfg = optarg;
        break;

      case 503:
        opt.config_avail = 1;
        config_avail = optarg;
        break;

      case 504:
        opt.config_need = 1;
        config_need = optarg;
        break;

      case 505:
        opt.new = 1;
        break;

      case 506:
        opt.fast = 1;
        break;

      case 507:
        opt.silent = 1;
        break;

      case 508:
        opt.boot = 1;
        break;

      case 509:
        opt.config_active = 1;
        config_active = optarg;
        break;

      case 510:
        if(*optarg) add_str_list(&opt.only, optarg);
        break;

      case 1000 ... 1100:
        opt.scan = 1;
        if(scan_items + 1 < sizeof scan_item / sizeof *scan_item) {
          scan_item[scan_items++] = i - 1000;
        }
        break;

      default:
        help();
        return 1;
    }
  }

  scan_item[scan_items] = 0;

  if(opt.scan && !opt.list) {
    if(argv[optind] || !scan_items) return help(), 1;
    rc = do_scan(scan_item);
    if(found_items) {
      unlink(HARDWARE_DIR "/.update");		/* the old file */
      unlink(HARDWARE_UNIQUE_KEYS "/.update");	/* so we trigger a rescan */
      if((f = fopen(HARDWARE_UNIQUE_KEYS "/.update", "a"))) fclose(f);
    }
    ok = 1;
  }

  if(opt.show) {
    do_show(id);
    ok = 1;
  }

  if(opt.list) {
    do_list(scan_item);
    ok = 1;
  }

  if(opt.config_cfg) {
    if(!argv[optind]) return help(), 1;
    do_config(1, config_cfg, argv[optind]);
    ok = 1;
  }

  if(opt.config_avail) {
    if(!argv[optind]) return help(), 1;
    do_config(2, config_avail, argv[optind]);
    ok = 1;
  }

  if(opt.config_need) {
    if(!argv[optind]) return help(), 1;
    do_config(3, config_need, argv[optind]);
    ok = 1;
  }

  if(opt.config_active) {
    if(!argv[optind]) return help(), 1;
    do_config(4, config_active, argv[optind]);
    ok = 1;
  }

  if(!ok) help();

#endif		/* !defined(LIBHD_TINY) */

  return rc;
}

void help()
{
  fprintf(stderr,
    "Usage: hwscan [options]\n"
    "Show information about currently known hardware.\n"
    "  --list            show list of known hardware\n"
    "  --version         show libhd version\n"
    "  --silent          don't show hardware config changes\n"
    "  --boot            run only if we haven't been disabled via 'hwprobe=-scan'\n"
    "  --cfg=state id    change 'configured' status; id is one of the\n"
    "                    ids from 'hwscan --list'\n"
    "                    state is one of new, no, yes\n"
    "  --avail=state id  change 'available' status\n"
    "  --need=state id   change 'needed' status\n"
    "  --active=state id change 'active' status\n"
    "  --hw_item         probe for hw_item and update status info\n"
    "  hw_item is one of:\n"
    "    cdrom, floppy, disk, network, gfxcard, framebuffer, monitor, camera,\n"
    "    mouse, joystick, keyboard, chipcard, sound, isdn, modem, storage-ctrl,\n"
    "    netcard, printer, tv, dvb, scanner, braille, sys, bios, cpu, partition,\n"
    "    usb-ctrl, usb, pci, isapnp, ide, scsi, bridge, hub, memory, smp, pppoe,\n"
    "    pcmcia, pcmcia-ctrl, wlan, dsl, zip\n"
  );
}

#ifndef LIBHD_TINY

int do_scan(hd_hw_item_t *items)
{
  int run_config = 0;
  hd_status_t status = { };
  hd_data_t *hd_data;
  hd_t *hd, *hd1;
  int err = 0;

  if(opt.fast) opt.fast = fast_ok(items);

  hd_data = calloc(1, sizeof *hd_data);

  if(opt.boot) {
    /* look if we have been disabled */
    hd_clear_probe_feature(hd_data, pr_all);
    hd_scan(hd_data);
    hd_set_probe_feature(hd_data, pr_scan);
    if(!hd_probe_feature(hd_data, pr_scan)) {
      hd_free_hd_data(hd_data);
      free(hd_data);
      return 0;
    }
  }

  hd_data->only = opt.only;
  opt.only = NULL;

  hd_data->flags.list_all = 1;
  hd_data->flags.fast = opt.fast;

  hd = hd_list2(hd_data, items, 1);

  if(hd) found_items = 1;

  for(hd1 = hd; hd1; hd1 = hd1->next) {
    err = hd_write_config(hd_data, hd1);
    if(verbose >= 2) {
      printf(
        "write=%d %s: (cfg=%s, avail=%s, need=%s, active=%s",
        err,
        hd1->unique_id,
        hd_status_value_name(hd1->status.configured),
        hd_status_value_name(hd1->status.available),
        hd_status_value_name(hd1->status.needed),
        hd_status_value_name(hd1->status.active)
      );
      if(hd1->unix_dev_name) {
        printf(", dev=%s", hd1->unix_dev_name);
      }
      printf(
        ") %s\n",
        hd1->model
      );
      
    }
    if(err) break;
  }

  if(err) {
    fprintf(stderr,
      "Error writing configuration for %s (%s)\n",
      hd1->unique_id,
      hd1->model
    );
    exit(1);
  }

  hd = hd_free_hd_list(hd);

  if(opt.new) {
    status.configured = status_new;
  }
  else {
    status.reconfig = status_yes;
  }

  hd = hd_list_with_status2(hd_data, items, status);
  if(hd) run_config = 1;

  if(verbose) {
    for(hd1 = hd; hd1; hd1 = hd1->next) {
      printf(
        "%s: (cfg=%s, avail=%s, need=%s, active=%s",
        hd1->unique_id,
        hd_status_value_name(hd1->status.configured),
        hd_status_value_name(hd1->status.available),
        hd_status_value_name(hd1->status.needed),
        hd_status_value_name(hd1->status.active)
      );
      if(hd1->unix_dev_name) {
        printf(", dev=%s", hd1->unix_dev_name);
      }
      printf(
        ") %s\n",
        hd1->model
      );
    }
  }
  else if(!opt.silent) {
    for(hd1 = hd; hd1; hd1 = hd1->next) printf("%s\n", hd1->unique_id);
  }

  hd = hd_free_hd_list(hd);

  hd_free_hd_data(hd_data);
  free(hd_data);

  return run_config ^ 1;
}


int do_show(char *id)
{
  hd_data_t *hd_data;
  hd_t *hd;

  hd_data = calloc(1, sizeof *hd_data);

  if ( id[0] == '/' ){
     int nr=0;
     char *_id = 0;
     hd_t *hd_manual;

     hd_manual = hd_list(hd_data, hw_manual, 1, NULL);
     for(hd = hd_manual; hd; hd = hd->next) {
        if(hd->status.available != status_yes) continue;
        if(!search_str_list(hd->unix_dev_names, id)) continue;
	_id = hd->unique_id;
        nr++;
     }

     if ( nr == 1 ) /* > 1 means our database is not okay */
        hd = hd_read_config(hd_data, _id);
  }else
    hd = hd_read_config(hd_data, id);

  if(hd) {
    hd_data->debug = -1;
    hd_dump_entry(hd_data, hd, stdout);
    hd = hd_free_hd_list(hd);
  }
  else {
    printf("no such hardware item: %s\n", id);
  }

  hd_free_hd_data(hd_data);
  free(hd_data);

  return 0;
}


int do_list(hd_hw_item_t *items)
{
  hd_data_t *hd_data;
  hd_t *hd, *hd_manual;
  char *s;
  char status[64];
  int i;

  hd_data = calloc(1, sizeof *hd_data);

  hd_manual = hd_list(hd_data, hw_manual, 1, NULL);

  for(hd = hd_manual; hd; hd = hd->next) {
    if(opt.scan && ! has_hw_class(hd, items)) continue;

    strcpy(status, "(");

    i = 0;
    if(hd->status.configured && (s = hd_status_value_name(hd->status.configured))) {
      sprintf(status + strlen(status), "%scfg=%s", i ? ", " : "", s);
      i++;
    }

    if(hd->status.available && (s = hd_status_value_name(hd->status.available))) {
      sprintf(status + strlen(status), "%savail=%s", i ? ", " : "", s);
      i++;
    }

    if(hd->status.needed && (s = hd_status_value_name(hd->status.needed))) {
      sprintf(status + strlen(status), "%sneed=%s", i ? ", " : "", s);
      i++;
    }

    if(hd->status.active && (s = hd_status_value_name(hd->status.active))) {
      sprintf(status + strlen(status), "%sactive=%s", i ? ", " : "", s);
      i++;
    }

    strcat(status, ")");

    s = hd_hw_item_name(hd->hw_class);
    if(!s) s = "???";

    printf("%s: %-32s %-16s %s\n", hd->unique_id, status, s, hd->model);
    if(hd->config_string) {
      printf("   configured as: \"%s\"\n", hd->config_string);
    }
  }

  hd_free_hd_list(hd_manual);

  hd_free_hd_data(hd_data);
  free(hd_data);

  return 0;
}


int do_config(int type, char *val, char *id)
{
  hd_data_t *hd_data;
  hd_t *hd;
  hd_status_value_t status = 0;
  int i;
  char *s;

  hd_data = calloc(1, sizeof *hd_data);

  if ( id[0] == '/' ){
     int nr=0;
     char *_id = 0;
     hd_t *hd_manual;

     hd_manual = hd_list(hd_data, hw_manual, 1, NULL);
     for(hd = hd_manual; hd; hd = hd->next) {
        if(hd->status.available != status_yes) continue;
        if(!search_str_list(hd->unix_dev_names, id)) continue;
	_id = hd->unique_id;
        nr++;
     }
     if ( nr == 1 )
        hd = hd_read_config(hd_data, _id);
  }else
    hd = hd_read_config(hd_data, id);

  if(hd) {
    for(i = 1; i < 8; i++) {
      s = hd_status_value_name(i);
      if(s && !strcmp(val, s)) {
        status = i;
        break;
      }
    }
    if(!status) {
      printf("invalid status: %s\n", val);
    }
    else {
      switch(type) {
        case 1:
          hd->status.configured = status;
          break;

        case 2:
          hd->status.available = status;
          break;

        case 3:
          hd->status.needed = status;
          break;

        case 4:
          hd->status.active = status;
          break;
      }
      hd_write_config(hd_data, hd);
    }
    hd = hd_free_hd_list(hd);
  }
  else {
    printf("no such hardware item: %s\n", id);
  }

  hd_free_hd_data(hd_data);
  free(hd_data);

  return 0;
}


/*
 * Check whether a 'fast' scan would suffice to re-check the presence
 * of all known hardware.
 */
int fast_ok(hd_hw_item_t *items)
{
  hd_data_t *hd_data;
  hd_t *hd, *hd1;
  int ok = 1;

  if(!has_item(items, hw_mouse) && !has_item(items, hw_storage_ctrl)) {
    return 1;
  }

  hd_data = calloc(1, sizeof *hd_data);

  hd_data->flags.list_all = 1;

  hd = hd_list(hd_data, hw_manual, 1, NULL);

  for(hd1 = hd; hd1; hd1 = hd1->next) {
    /* serial mice */
    if(hd1->hw_class == hw_mouse && hd1->bus.id == bus_serial) {
      ok = 0;
      break;
    }
    /* parallel zip */
    if(hd1->hw_class == hw_storage_ctrl && hd1->bus.id == bus_parallel) {
      ok = 0;
      break;
    }
  }

  hd_free_hd_data(hd_data);
  free(hd_data);

  return ok;
}


/* check if item is in items */
int has_item(hd_hw_item_t *items, hd_hw_item_t item)
{
  while(*items) if(*items++ == item) return 1;

  return 0;
}


/* check if one of items is in hw_class */
int has_hw_class(hd_t *hd, hd_hw_item_t *items)
{
  while(*items) if(hd_is_hw_class(hd, *items++)) return 1;

  return 0;
}


#endif		/* !defined(LIBHD_TINY) */
